File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""
A real rectangle module
"""


class Rectangle:
    """
    A class in python3 that is
    Inclusive of the area and perimeter
    """

    def __init__(self, width=0, height=0):
        """Instantiate"""
        self.width = width
        self.height = height

    @property
    def width(self):
        return (self.__width__)

    @width.setter
    def width(self, value):
        if not isinstance(value, int):
            raise TypeError("width must be an integer")
        elif (value < 0):
            raise ValueError("width must be >= 0")
        else:
            self.__width__ = value

    @property
    def height(self):
        return (self.__height__)

    @height.setter
    def height(self, value):
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        elif (value < 0):
            raise ValueError("height must be >= 0")
        else:
            self.__height__ = value

    def area(self):
        return (self.width * self.height)

    def perimeter(self):
        if (self.width == 0) or (self.height == 0):
            return (0)
        return 2 * (self.width + self.height)

    def print(self):
        if (self.width == 0) or (self.height == 0):
            return("")
        else:
            for i in range(self.width):
                for j in range(self.height):
                    print("#", end="")
                print()

    def __str__(self):
        if (self.width == 0) or (self.height == 0):
            return("")
        else:
            for i in range(self.height):
                for j in range(self.width):
                    print("#", end="")
                print()
            return ("")

    def __repr__(self):
        return ("Rectangle(" + 
            str(self.width) + ", " + 
            str(self.height) + ")")

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let string = '';
    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        string += 'X';
      }
      console.log(string);
      string = '';
    }
  }

  rotate () {
    const temp = this.height;
    this.height = this.width;
    this.width = temp;
  }

  double () {
    this.width *= 2;
    this.height *= 2;
  }
}

module.exports = Rectangle;
